### PR TITLE
feat:  preventing the customers to create a patient with a DOB in the future

### PR DESCRIPTION
**Pull request by husainsyed**
Ticket: [1853](https://github.com/metriport/metriport-internal/issues/1853)

### Description
- Preventing the customers from creating patients with a DOB in the future:
- We're replacing `validateIsPastOrPresent` with `validateDateOfBirth` as we now ensure our patient info is valid (>= 1900 as well as not a future dob)

### Testing
- Local
  - Unit tests

### Release Plan

- [ ] Merge this

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved validation for patient date of birth to ensure only valid and correctly formatted dates after 1900 are accepted.
- **Tests**
	- Expanded and refined test coverage for date validation, including additional scenarios and stricter checks for date formats and boundaries.
	- Updated test for future date of birth validation with extended future offset.
- **Chores**
	- Updated internal references to external subprojects.

### CODE REVIEW

File: packages/api/src/command/medical/patient/shared.ts [hunk 1]
```diff
@@ -18,7 +22,7 @@ export function validate<T extends PatientCreateCmd | PatientUpdateCmd | Patient
 ): boolean {
   if (!patient.address || patient.address.length < 1) return false;
   patient.personalIdentifiers?.forEach(pid => pid.period && validatePeriod(pid.period));
-  validateIsPastOrPresent(patient.dob);
+  validateDateOfBirth(patient.dob);
```
[the remark this answers is not in the document]
Comment (husainsyed) [reply packages/api/src/command/medical/patient/shared.ts:25]: good catch, let me add that.

File: packages/shared/src/common/__tests__/date.test.ts [hunk 2]
```diff
@@ -40,4 +40,12 @@ describe("validateDateIsAfter1900", () => {
   it("returns true for 1900-01-01", () => {
     expect(validateDateIsAfter1900("1900-01-01")).toBe(true);
   });
+
+  it("returns false for 0007-01-01", () => {
+    expect(validateDateIsAfter1900("0007-01-01")).toBe(false);
+  });
+
+  it("returns false for 0014-01-01", () => {
```
Comment (RamilGaripov) [inline packages/shared/src/common/__tests__/date.test.ts:48]: I think it'd be good to be less specific in the description for these tests. Something like:
`returns false if year is between 0 and 9 year AD`
`returns false if year is between 10 and 99 year AD`

**Review by RamilGaripov — CHANGES_REQUESTED** [review]

File: packages/api/src/command/medical/patient/shared.ts [hunk 3]
```diff
@@ -18,7 +22,10 @@ export function validate<T extends PatientCreateCmd | PatientUpdateCmd | Patient
 ): boolean {
   if (!patient.address || patient.address.length < 1) return false;
   patient.personalIdentifiers?.forEach(pid => pid.period && validatePeriod(pid.period));
-  validateIsPastOrPresent(patient.dob);
+
+  if (patient.dob) {
```
Comment (leite08) [inline packages/api/src/command/medical/patient/shared.ts:26]: Can `dob` be optional here?

File: packages/api/src/command/medical/patient/__tests__/validate.test.ts [hunk 4]
```diff
@@ -46,8 +46,8 @@ describe("validate", () => {
       expect(() => validate(patient)).toThrow(`Date can't be in the future`);
     });
 
-    it("throws an error when dob is a minute in the future", async () => {
-      const futureDate = buildDayjs().add(1, "minute").toISOString();
+    it("throws an error when dob is an hour in the future", async () => {
```
[the remark this answers is not in the document]
Comment (husainsyed) [reply packages/api/src/command/medical/patient/__tests__/validate.test.ts:49]: Looked into the issue and found that some times the CI environments conflict w/ the local code times. Meaning: Since the test case tested for "1 minute in the future" before, by the time the code execution happens, 1 minute has already passed so it's failing the test case... so I changed it to "1 hour in the future". But obv that wasn't the problem either.

File: packages/shared/src/common/__tests__/date.test.ts (hunk 2 above)
Comment (husainsyed) [reply packages/shared/src/common/__tests__/date.test.ts:48]: I guess I wasn't clear... the current test in the GitHub CI returns:
```
FAIL src/command/medical/patient/__tests__/validate.test.ts
  ● validate › dob › throws an error when dob is an hour in the future

    expect(received).toThrow(expected)

    Expected substring: "Date can't be in the future"

    Received function did not throw
```

But in the local, the function IS throwing `"Date can't be in the future"`

--

Once I figure out whats the problem, I'll make the message more descriptive :)

File: packages/api/src/command/medical/patient/__tests__/validate.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (husainsyed) [reply packages/api/src/command/medical/patient/__tests__/validate.test.ts:49]: the test serves the same purpose.. be in a minute or hour in the future.. thats why didn't make the change in the new commit.

File: packages/api/src/command/medical/patient/shared.ts (hunk 3 above)
Comment (husainsyed) [reply packages/api/src/command/medical/patient/shared.ts]: addressed in the latest push

File: packages/api/src/command/medical/patient/__tests__/validate.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/command/medical/patient/__tests__/validate.test.ts:49]: > the test serves the same purpose 

that's why i was asking. If it serves the same purpose, then changing it is a hack 😅 
i dont mind it staying the way it is now, but in the future, i think it's faster to revert something like this than having to explain why it was changed and have the reviewer also read and think about it. But NBAD :)

File: packages/shared/src/common/__tests__/date.test.ts [hunk 5]
```diff
@@ -45,6 +45,11 @@ describe("validateDateIsAfter1900", () => {
   it("returns true for 1900-01-01", () => {
     expect(validateDateIsAfter1900("1900-01-01")).toBe(true);
   });
+
+  it("returns false for dates with years less than 1000", () => {
```
Comment (RamilGaripov) [inline packages/shared/src/common/__tests__/date.test.ts:49]: with a descriptor like this, i'd also expect a test case for `0xxx-xx-xx`

Comment (husainsyed) [reply packages/shared/src/common/__tests__/date.test.ts:80]: done

File: packages/shared/package.json [hunk 6]
```diff
@@ -80,11 +80,13 @@
   },
   "dependencies": {
     "@medplum/core": "^2.2.10",
+    "@types/luxon": "^3.6.2",
     "axios": "^1.4.0",
     "dayjs": "^1.11.9",
     "fast-xml-parser": "^4.4.1",
     "http-status": "~1.7.0",
     "lodash": "^4.17.21",
+    "luxon": "^3.6.1",
```
[the remark this answers is not in the document]
Comment (husainsyed) [reply packages/shared/package.json:89]: I do not intend to migrate all the instances of Day.js. This ISO thing was an anomaly - the Day.js library doesn't handle ISO dates well. The other option was regex that would get very messy, very quickly. So just for the ISO validation, I opted to use luxon.

**Comment by husainsyed** [comment]
@RamilGaripov // @leite08 
I have combined the changes from [PR 3661](https://github.com/metriport/metriport/pull/3661) into this PR because of overlapping code. That PR can safely be destroyed.

Since both PR didn't share any commit history, I couldn't establish any upstream dependency. This was the safest & fastest option w/o getting into the nitty gritty of git commit history and rebasing.

File: package.json [hunk 7]
```diff
@@ -82,6 +82,7 @@
     "typescript": "^4.9.5"
   },
   "dependencies": {
+    "@types/luxon": "^3.6.2",
```
Comment (leite08) [inline package.json:85]: We shouldn't add dependencies on the root package.json. The "sharp" one is an anomaly and eventually should be removed.

File: packages/shared/package.json [hunk 8]
```diff
@@ -80,11 +80,13 @@
   },
   "dependencies": {
     "@medplum/core": "^2.2.10",
+    "@types/luxon": "^3.6.2",
```
Comment (RamilGaripov) [inline packages/shared/package.json:83]: `@types` go under `devDependencies`

File: packages/api/src/command/medical/patient/shared.ts [hunk 9]
```diff
@@ -18,8 +22,7 @@ export function validate<T extends PatientCreateCmd | PatientUpdateCmd | Patient
 ): boolean {
   if (!patient.address || patient.address.length < 1) return false;
   patient.personalIdentifiers?.forEach(pid => pid.period && validatePeriod(pid.period));
-  validateIsPastOrPresent(patient.dob);
-  return true;
+  return !patient.dob || (validateIsPastOrPresent(patient.dob) && validateDateOfBirth(patient.dob));
```
Comment (leite08) [inline packages/api/src/command/medical/patient/shared.ts:25]: Why checking for the presence of `dob`, isn't it just `string`?

Why calling `validateIsPastOrPresent` if that's called in `validateDateOfBirth`?

File: packages/shared/src/common/__tests__/date.test.ts [hunk 10]
```diff
@@ -7,9 +7,40 @@ import {
...
 describe("shared date functions", () => {
   describe("isValidISODate", () => {
-    it("returns true for dates from E2E tests", async () => {
+    it("returns true for valid ISO datetime strings", () => {
       expect(isValidISODate("2024-12-18T03:50:00.006Z")).toEqual(true);
       expect(isValidISODate("2024-12-18T04:18:01.263Z")).toEqual(true);
+      expect(isValidISODate("2024-12-18T03:50:00.006Z")).toBe(true);
+      expect(isValidISODate("2024-12-18T04:18:01.263Z")).toBe(true);
+      expect(isValidISODate("2024-12-18T23:59:59.999Z")).toBe(true);
+    });
```
Comment (leite08) [inline packages/shared/src/common/__tests__/date.test.ts:16]: Let's avoid the pattern to have multiple tests on a single unit test. Those are supposed to be "unit" level, a single test case.

**Review by RamilGaripov — CHANGES_REQUESTED** [review]
just the one change in pkg.json. Once you fix it, rerun npm i to update pkg lock file

File: packages/shared/src/common/date.ts [hunk 11]
```diff
@@ -2,22 +2,30 @@ import dayjs, { ConfigType } from "dayjs";
...
 
 dayjs.extend(utc);
 
 export const ISO_DATE = "YYYY-MM-DD";
 
 /** @see https://day.js.org/docs/en/parse/is-valid  */
 export function isValidISODate(date: string): boolean {
-  return buildDayjs(date, ISO_DATE, true).isValid();
+  const isoPattern = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/;
+  return isoPattern.test(date) && DateTime.fromISO(date, { zone: "utc" }).isValid;
```
Comment (leite08) [inline packages/shared/src/common/date.ts:14]: Why are we replacing `dayjs.isValid`?

Also, please flag and explain why you propose a new library, we avoid adding additional dependencies and need to evaluate carefully when doing so.

Note: this function is not to convert any date into an ISO date, it's just to validate if the date is in this format - returning false if it's not.

TBH, I don't even remember why we're touching this function in the first place. 😄 IIRC, I asked about adding test cases for the `validateDateIsAfter1900` with input date non-ISO.

File: packages/shared/package.json (hunk 8 above)
Comment (husainsyed) [reply packages/shared/package.json]: done

File: packages/shared/src/common/date.ts (hunk 11 above)
Comment (husainsyed) [reply packages/shared/src/common/date.ts:14]: 1. dayjs doesn't handle strict non-iso cases. The existing logic failed most of the tests I added. The new dependency takes care of all the non-iso stuff.

2. `Can we add tests for dates that are not in ISO format, including empty string, please?` - I assumed this is for the ISO validation.

Let me know how you want me to proceed.

File: packages/api/src/command/medical/patient/shared.ts (hunk 9 above)
Comment (husainsyed) [reply packages/api/src/command/medical/patient/shared.ts:25]: 1. Early return basically. If `!patient.dob` is true (meaning dob is not provided), it doesn't get into extra validation.
2. Removing the method fails the `const futureDate = buildDayjs().add(1, "hour").toISOString();` test case because:

`validateIsPastOrPresent(patient.dob)` --> checks time precision (hours/minutes)
`validateDateOfBirth` --> checks only date precision (year/month/day)

it loses precision for dates that are on the same day but  in the future

File: package.json (hunk 7 above)
Comment (husainsyed) [reply package.json:85]: done

File: packages/shared/src/common/__tests__/date.test.ts (hunk 10 above)
Comment (husainsyed) [reply packages/shared/src/common/__tests__/date.test.ts]: Done.

File: packages/api/src/command/medical/patient/shared.ts (hunk 9 above)
Comment (RamilGaripov) [reply packages/api/src/command/medical/patient/shared.ts:25]: What Raf means is that `dob` is just `string` and is not `string | undefined`. So why do we need to check if it's present? It should already be always present...

File: packages/shared/src/common/date.ts (hunk 11 above)
Comment (RamilGaripov) [reply packages/shared/src/common/date.ts:14]: What's the purpose of these tests? What are we trying to achieve? 
It's one thing to write tests that expect a failed validation, and another thing to change the logic to make every date format to be valid. 

AFAIK, we can continue to expect our CX to provide the DOB in the existing format. We just needed to add tests to make sure that other format are, in fact, unsupported

File: packages/api/src/command/medical/patient/shared.ts (hunk 9 above)
Comment (husainsyed) [reply packages/api/src/command/medical/patient/shared.ts]: 1. Addressed ✅

2. What I have done is I am now calling `validateDateIsAfter1900` in `validateDateOfBirth`.
`validate` has `validateIsPastOrPresent(patient.dob) && validateDateOfBirth(patient.dob);`  Tried every other way -- the error message gets swallowed by the layered method calls, ends up failing the test case testing the `future dates`

File: packages/shared/src/common/date.ts (hunk 11 above)
Comment (husainsyed) [reply packages/shared/src/common/date.ts]: Removed the `luxon` dependency.
Restored the `isValidISODate` method along w/ additional tests for `validateDateIsAfter1900`.

The `validateDateIsAfter1900` needed some extra modifications so addressed those as well.

File: packages/api/src/command/medical/patient/shared.ts [hunk 12]
```diff
@@ -18,8 +22,7 @@ export function validate<T extends PatientCreateCmd | PatientUpdateCmd | Patient
 ): boolean {
   if (!patient.address || patient.address.length < 1) return false;
   patient.personalIdentifiers?.forEach(pid => pid.period && validatePeriod(pid.period));
-  validateIsPastOrPresent(patient.dob);
-  return true;
+  return validateIsPastOrPresent(patient.dob) && validateDateOfBirth(patient.dob);
```
Comment (RamilGaripov) [inline packages/api/src/command/medical/patient/shared.ts:25]: If we need to perform both checks in order to validate dob, then it seems like validateDateOfBirth doesn't do what its name indicates. I insist that we write that function in a way that makes it fully sufficient to tell whether the date is valid or not

**Review by RamilGaripov — DISMISSED** [review]

**Review by RamilGaripov — APPROVED** [review]
Approving for now, but expecting the comments around validateDateOfBirth to be addressed soon
